src/pke/lib/scheme/ckksrns/ckksrns-multiparty.cpp

```diff
@@ -98,4 +115,347 @@ DecryptResult MultipartyCKKSRNS::MultipartyDecryptFusion(const std::vector<Ciphe
+    }
+
+    // compute the [Q/q_i]_{p_j}
+    // used for homomorphic multiplication
+    rnsExtTables.QHatModp.resize(sizeP);
+    for (usint j = 0; j < sizeP; j++) {
+        BigInteger pj(moduliP[j].ConvertToInt());
+        for (usint i = 0; i < sizeQ; i++) {
+            BigInteger qi(moduliQ[i].ConvertToInt());
+            BigInteger QHati = modulusQ / qi;
+            rnsExtTables.QHatModp[j].push_back(QHati.Mod(pj).ConvertToInt());
+        }
+    }
+
+    // compute the [\alpha*Q]p_j for 0 <= alpha <= sizeQ
+    // used for homomorphic multiplication
+    rnsExtTables.alphaQModp.resize(sizeQ + 1);
+    for (usint j = 0; j < sizeP; j++) {
+        BigInteger pj(moduliP[j].ConvertToInt());
+        NativeInteger QModpj = modulusQ.Mod(pj).ConvertToInt();
+        for (usint i = 0; i < sizeQ + 1; i++) {
+            rnsExtTables.alphaQModp[i].push_back(QModpj.ModMul(NativeInteger(i), moduliP[j]));
+        }
+    }
+
+    // Precomputations for Barrett modulo reduction
+    const BigInteger BarrettBase128Bit("340282366920938463463374607431768211456");  // 2^128
+    const BigInteger TwoPower64("18446744073709551616");                            // 2^64
+    rnsExtTables.modpBarrettMu.resize(sizeP);
+    for (uint32_t j = 0; j < moduliP.size(); j++) {
+        BigInteger mu = BarrettBase128Bit / BigInteger(moduliP[j]);
+        uint64_t val[2];
+        val[0] = (mu % TwoPower64).ConvertToInt();
+        val[1] = mu.RShift(64).ConvertToInt();
+
+        memcpy(&rnsExtTables.modpBarrettMu[j], val, sizeof(DoubleNativeInt));
+    }
+
+    rnsExtTables.qInv.resize(sizeQ);
+    for (size_t i = 0; i < sizeQ; i++) {
+        rnsExtTables.qInv[i] = 1. / static_cast<double>(moduliQ[i].ConvertToInt());
+    }
+}
+
+// Utility function to compute noisy multiplication ( sk * poly + noise )
+// noise will not be added if IsZeroNoise is set to true (as in computing h_0,i)
+DCRTPoly ComputeNoisyMult(CryptoContext<DCRTPoly>& cc, const DCRTPoly& sk, const DCRTPoly& poly, bool IsZeroNoise) {
+    if (sk.GetNumOfElements() != poly.GetNumOfElements()) {
+        std::string errMsg = "ERROR: Number of towers in input polys does not match!";
+        OPENFHE_THROW(config_error, errMsg);
+    }
+
+    DCRTPoly res = sk * poly;
+    if (false == IsZeroNoise) {
+        const auto cryptoParams      = std::dynamic_pointer_cast<CryptoParametersRNS>(cc->GetCryptoParameters());
+        const DCRTPoly::DggType& dgg = cryptoParams->GetDiscreteGaussianGenerator();
+        auto paramsq                 = poly.GetParams();
+
+        DCRTPoly e(dgg, paramsq, Format::EVALUATION);
+        res = res + e;
+    }
+
+    return res;
+}
+
+// Generate random mask
+DCRTPoly GenerateMi(const DCRTPoly& c1, uint32_t maskBoundNumTowers) {
+    auto c1Copy = c1;
+
+    // drop twoers until we reach maskBoundNumTowers
+    while (c1Copy.GetAllElements().size() > maskBoundNumTowers)
+        c1Copy.DropLastElement();
```

Comment:
DropLastElements() should be used instead of DropLastElement()

Reply:
Fixed as suggested